core/src/main/scala/tapir/internal/server/DecodeInputs.scala

```diff
@@ -71,119 +73,204 @@ object DecodeInputs {
+    var decodeInputResult: DecodeInputsResult = origDecodeValues
+    var thunks = Vector.empty[() => (IndexedBasicInput, DecodeResult[_])]
+    nonEmptyInputs.foreach {
+      case indexedInput @ IndexedBasicInput(endpointInput, _) =>
+        decodeInputResult match {
+          case _: DecodeInputsResult.Failure => () // Already failed so just let the loop continue
+          case _: DecodeInputsResult.Values => {
+            val (nextSegmentOpt, nextCtx) = currCtx.nextPathSegment
+            currCtx = nextCtx
+            nextSegmentOpt match {
+              case Some(actualSegment) =>
+                endpointInput match {
+                  case EndpointInput.FixedPath(expectedSegment) =>
+                    if (actualSegment == expectedSegment) {
+                      ()
+                    } else {
+                      decodeInputResult = DecodeInputsResult.Failure(endpointInput, DecodeResult.Mismatch(expectedSegment, actualSegment))
+                    }
+                  case pathCaptureInput: EndpointInput.PathCapture[_] => {
+                    thunks = thunks :+ { () =>
+                      (indexedInput, pathCaptureInput.codec.safeDecode(actualSegment))
+                    }
+                  }
+                  case _: EndpointInput.PathsCapture =>
+                    val (restSegments, ctxNoMorePaths) = collectRemainingPath(Vector.empty, currCtx)
+                    currCtx = ctxNoMorePaths
+                    thunks = thunks :+ { () =>
+                      (indexedInput, DecodeResult.Value(actualSegment +: restSegments))
+                    }
+                  case _ =>
+                    throw new Error(
+                      s"Unexpected EndpointInput ${indexedInput.input.show} encountered. " +
+                        s"This is most likely a bug in the library"
+                    )
+
+                }
+              case None =>
+                endpointInput match {
+                  case EndpointInput.FixedPath(expectedPath) => {
+                    if (expectedPath.nonEmpty) {
+                      decodeInputResult = DecodeInputsResult.Failure(endpointInput, DecodeResult.Missing)
+                    } else {
+                      // FixedPath("") matches an empty path
+                      ()
+                    }
+                  }
+                  case _: EndpointInput.PathCapture[_] => {
+                    decodeInputResult = DecodeInputsResult.Failure(endpointInput, DecodeResult.Missing)
+                  }
+                  case _: EndpointInput.PathsCapture => {
+                    // Capture empty path
+                    decodeInputResult = decodeInputResult match {
+                      case failure: DecodeInputsResult.Failure => failure
+                      case values: DecodeInputsResult.Values   => values.setBasicInputValue(Vector.empty[String], indexedInput.index)
+                    }
+                  }
+                  case _ =>
```

Comment:
@adamw How do you feel about making all path related `EncodeInput`s extend a common sealed trait so I can avoid this wildcard match?

Reply:
I know it's tempting, but there's already a number of traits (Basic, Single, Input, IO), so I think I'd try to avoid more of those :)